raft timeouts: better handling of lost quorum

gusev-p opened the pull request:
In this PR we add timeouts support to raft groups registry. We introduce the `raft_server_with_timeouts` class, which wraps the `raft::server` add exposes its interface with additional `raft_timeout` parameter. If it's set, the wrapper cancels the `abort_source` after certain amount of time. The value of the timeout can be specified either in the `raft_timeout` parameter, or the default value can be set in `the raft_server_with_timeouts` class constructor.

The `raft_group_registry` interface is extended with `group0_with_timeouts()` method. It returns an instance of `raft_server_with_timeouts` for group0 raft server. The timeout value for it is configured in `create_server_for_group0`. It's one minute by default and can be overridden for tests with `group0-raft-op-timeout-in-ms` parameter.

The new api allows the client to decide whether to use timeouts or not. In this PR we are reviewing all the group0 call sites and add `raft_timeout` if that makes sense. The general principle is that if the code is handling a client request and the client expects a potential error, we use timeouts. We don't use timeouts for background fibers (such as topology coordinator), since they wouldn't add much value. The only thing the background fiber can do with a timeout is to retry, and this will have the same end effect as not having a timeout at all.

fixes scylladb/scylladb#16604

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7098/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7098/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/233/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7098/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8273/)
#### Failed Tests (4/23525):
- [test_quorum_lost_during_node_join_response_handler](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7098/testReport/junit/%28root%29/test_raft_no_quorum/test_quorum_lost_during_node_join_response_handler) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_quorum_lost_during_node_join_response_handler)
- [test_quorum_lost_during_node_join_response_handler](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7098/testReport/junit/%28root%29/test_raft_no_quorum/test_quorum_lost_during_node_join_response_handler_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_quorum_lost_during_node_join_response_handler)
- [topology_custom.test_raft_no_quorum.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7098/testReport/junit/%28root%29/non-boost%20tests/topology_custom_test_raft_no_quorum_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_custom.test_raft_no_quorum.debug.1)
- [topology_custom.test_raft_no_quorum.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7098/testReport/junit/%28root%29/non-boost%20tests/topology_custom_test_raft_no_quorum_debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_custom.test_raft_no_quorum.debug.2)
#### Build Details:
- Duration: 1 hr 58 min
- Builder: spider8.cloudius-systems.com

gusev-p commented:
new version:
  * increase timeouts in tests to make them pass on the CI

gusev-p commented:
@yaronkaikov Hi, how can I find out what's happened to the CI job? Only the spelling check managed to run.

yaronkaikov commented:
> @yaronkaikov Hi, how can I find out what's happened to the CI job? Only the spelling check managed to run.

It seems like a bug in the CI, looks like you pushed more then once in matter of minutes and it caused 
![image](https://github.com/scylladb/scylladb/assets/9914527/71ec2a98-67d4-4b21-94f2-5c084c9a9d19)


@benipeled FYI

I will open an issue to understand this issue, in the meantime , I re-triggered the CI - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7128/

yaronkaikov commented:
> > > @yaronkaikov Hi, how can I find out what's happened to the CI job? Only the spelling check managed to run.
> > 
> > 
> > It seems like a bug in the CI, looks like you pushed more then once in matter of minutes and it caused ![image](https://private-user-images.githubusercontent.com/9914527/309712046-71ec2a98-67d4-4b21-94f2-5c084c9a9d19.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDk1NDY4NjYsIm5iZiI6MTcwOTU0NjU2NiwicGF0aCI6Ii85OTE0NTI3LzMwOTcxMjA0Ni03MWVjMmE5OC02N2Q0LTRiMjEtOTRmMi01YzA4NGM5YTlkMTkucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDMwNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAzMDRUMTAwMjQ2WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NTM4NGVlOGJjMzEyZTk5YzE1NTM5OGY5YjdjNDg2YmUyZGFkYmYwNDY0NjRhMTAzMzUwMGMwZjU3MzJhMzljMCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.jvlisvnw5ybBW4daSjVOfgEmvsOl46_tMbew72brlhs)
> > @benipeled FYI
> > I will open an issue to understand this issue, in the meantime , I re-triggered the CI - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7128/
> 
> Thanks!

https://github.com/scylladb/scylla-pkg/issues/3875

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7128/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7128/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/272/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8310/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7128/)
#### Build Details:
- Duration: 1 hr 46 min
- Builder: spider8.cloudius-systems.com

patjed41 commented on service/raft/raft_group_registry.cc:
```diff
@@ -408,6 +413,98 @@ shared_ptr<raft::failure_detector> raft_group_registry::failure_detector() {
+        }
+        timeout->value = _timeout.value();
+    }
+    composite_abort_source composite_as;
+
+    abort_on_expiry<> expiry{lowres_clock::now() + *timeout->value};
+    composite_as.add(expiry.abort_source());
+
+    if (as) {
+        composite_as.add(*as);
+    }
+
+    try {
+        co_return co_await op(&composite_as.abort_source());
```
Is the `composite_abort_source` class really necessary? Can't we just do something like:
```cpp
abort_on_expiry<> expiry{lowres_clock::now() + *timeout->value};
auto sub = expiry.abort_source().subscribe([as]() noexcept { if (as) as->request_abort(); });

try {
    co_return co_await op(as ? *as : expiry.abort_source());
```

gusev-p replied:
`subscribe([as]() noexcept { if (as) as->request_abort(); });`
this is incorrect, we need to trigger abort no matter if we have `as` or not.

We could subscribe to `as` and trigger `expiry.abort_source()`, but that would be confusing.

It's better to extract the wiring of `abort_source`-s, it can be reused elsewhere.

patjed41 replied:
> `subscribe([as]() noexcept { if (as) as->request_abort(); });` this is incorrect, we need to trigger abort no matter if we have `as` or not.

This change
```cpp
co_return co_await op(as ? as : &expiry.abort_source());
```
should ensure that if we don't have `as` and we time out, we will trigger abort.

I found a different problem with my idea though. We cannot distinguish only `expiry.abort_source()` triggering from both `expiry.abort_source()` and `as` triggering. We would have to treat the second case as a timeout, which is different from the current code. The if below would change to
```cpp
if (!expiry.abort_source().abort_requested()) {
    throw;
}
```
I don't think it really matters. Both abort sources triggering seems like an extremely rare case and we don't know what was triggered first anyway. 

> It's better to extract the wiring of `abort_source`-s, it can be reused elsewhere.

Yes, I don't mind it. It just seemed to me like a bit of an overkill for this particular problem.

gusev-p replied:
> This change
> 
> ```c++
> co_return co_await op(as ? as : &expiry.abort_source());
> ```
> 
> should ensure that if we don't have `as` and we time out, we will trigger abort.

We can't trigger the `as` from parameters in case of timeout, we don't own it, some unrelated code can depend on it.

If you want to save on `abort_source`-s,  you can pass `expiry.abort_source()` to `co_await op()` and trigger it through `as.subscribe`, but that would be confusing, as I said above. We would need to change the the `if` below
```
if (!expiry.abort_source().abort_requested() || (as && as->abort_requested())) {
    throw;
}
```
With the current structure each `abort_source` has its own responsibility and the `if` is clear.

patjed41 replied:
> We can't trigger the `as` from parameters in case of timeout, we don't own it, some unrelated code can depend on it.

Right.

Thank you for the explanation.

patjed41 commented on test/topology_custom/test_raft_no_quorum.py:
```diff
@@ -54,7 +58,7 @@ async def test_cannot_add_new_node(manager: ManagerClient) -> None:
 
     logger.info("starting a fifth node with no quorum")
-    await manager.server_add(expected_error="raft operation [read_barrier] timed out",
+    await manager.server_add(expected_error="raft operation [read_barrier] timed out, no group0 quorum",
```
In the last commit you additionally match "no group0 quorum, total voters 2, alive voters 1". Maybe you could also match it here and in other tests?

gusev-p replied:
I wanted to check somewhere the logic which calculates alive voters. I did it once. The focus of other tests is on the specific scenarios where this "no group0 quorum" can happen, so I check just that. We could add the specific numbers of voters, but that would be redundant.

patjed41 commented on test/pylib/manager_client.py:
```diff
@@ -279,7 +283,7 @@ async def servers_add(self, servers_num: int = 1,
         assert servers_num > 0, f"servers_add: cannot add {servers_num} servers, servers_num must be positive"
 
         try:
-            data = self._create_server_add_data(None, cmdline, config, property_file, start, expected_error)
+            data = self._create_server_add_data(None, cmdline, config, property_file, start, None, expected_error)
```
Additionally adding `seeds` to `servers_add` is trivial and might be useful later (especially for someone who doesn't know test.py internals and won't extend the API easily). This PR doesn't need this change though, it's optional.

gusev-p replied:
Yes, can do that if we need to change something else, would like to avoid rewriting git/rebuilding/CI run only for that.

patjed41 commented on service/storage_service.cc:
```diff
@@ -4351,7 +4351,7 @@ future<> storage_service::raft_check_and_repair_cdc_streams() {
 
```
Is there a reason why you didn't add a test for `cdc_streams_check_and_repair` and other REST api operations? Is it hard to test or not needed?

gusev-p replied:
There are a lot of places where this PR adds timeouts, adding a test for each of them would add too much CI overhead with little value - the changes are trivial.

gusev-p commented on service/raft/raft_group_registry.cc:
```diff
@@ -289,8 +289,11 @@ seastar::future<> raft_group_registry::start() {
 
-    _direct_fd_subscription.emplace(co_await _direct_fd.register_listener(*_direct_fd_proxy,
-        direct_fd_clock::base::duration{std::chrono::seconds{1}}.count()));
+    direct_fd_clock::base::duration threshold{std::chrono::seconds{1}};
+    if (const auto ms = utils::get_local_injector().inject_parameter<int64_t>("raft-group-registry-fd-threshold-in-ms"); ms) {
+        threshold = direct_fd_clock::base::duration{std::chrono::milliseconds{*ms}};
+    }
+    _direct_fd_subscription.emplace(co_await _direct_fd.register_listener(*_direct_fd_proxy, threshold.count()));
```
It's better to declare assumptions explicitly in tests. With this injection the tests have both knobs at their disposal and can request one to be smaller than the other. Without this injection we would have to implicitly rely on the exact hardcoded timeout value, and the test will break if that timeouts change.

kbr-scylla commented on service/raft/raft_group_registry.hh:
```diff
@@ -49,6 +71,7 @@ struct raft_server_for_group {
     raft_rpc& rpc;
     raft_sys_table_storage& persistence;
     std::optional<seastar::future<>> aborted;
+    std::optional<lowres_clock::duration> timeout;
```
I'm a bit confused, there's 2 classes now that store a timeout and then there's the timeout parameter passed to API functions.

I wonder if it wouldn't be simpler to have `raft_server_with_timeouts` **own** `raft::server`, and then `raft_server_for_group` would hold a `unique_ptr` to `raft_server_with_timoeuts`.

We could implement `raft_server_with_timeout` through private inheritance from `raft::server`. And then it would selectively expose the interface of `raft::server` (functions like `stepdown`) vs its own (the timeout variants of certain functions like `add_entry`).

Just something to consider.

kbr-scylla replied:
> Why inherit privately from raft::server?

Does the hierarchy satisfy LSP? Maybe it does (didn't think too much about it), in which case public would be OK too.

kbr-scylla commented on service/raft/raft_group0_client.cc:
```diff
@@ -153,7 +153,9 @@ semaphore& raft_group0_client::operation_mutex() {
 
-future<> raft_group0_client::add_entry(group0_command group0_cmd, group0_guard guard, seastar::abort_source* as) {
+future<> raft_group0_client::add_entry(group0_command group0_cmd, group0_guard guard, seastar::abort_source* as,
+        std::optional<raft_timeout> timeout)
```
I don't understand why you made these design choices.

Some callers of `raft_group0_client::add_entry` -- those that want to provide a `timeout` -- will provide both `as` and `timeout`.

So why not do the "composite abort_source" thing on the caller of `raft_group0_client::add_entry`? Why  introduce abstraction inside `raft_server_for_group`?

This way the interface of `raft_group0_client::add_entry` would still be simple (just taking `abort_source*`); if the caller wants timeout, they would pass a composite abort_source that includes a timeout source.

(etc. for other places)

kbr-scylla replied:
After reading the rest of the PR, it makes sense, especially with the extra reporting of quorum loss etc. (you'd basically have to wrap each raft::server API call with something that would do the reporting)

kbr-scylla commented on test/topology_custom/test_raft_no_quorum.py:
```diff
@@ -37,6 +37,10 @@ async def test_cannot_add_new_node(manager: ManagerClient) -> None:
+            },
+            {
+                'name': 'raft-group-registry-fd-threshold-in-ms',
+                'value': '500'
```
Hopefully the test won't be flaky. Let's run it 100 times in debug mode on CI (using the byo_build_test_dtest job)

kbr-scylla commented on service/raft/raft_group_registry.cc:
```diff
@@ -473,8 +476,24 @@ raft_server_with_timeouts::run_with_timeout(Op&& op, const char* op_name,
+                }
+            }
+            if (alive_voters_count <= voters_count / 2) {
+                message += ::format(", no group0 quorum, total voters {}, alive voters {}",
```
Nice.

kbr-scylla replied:
Although it would be better (performance-wise) if we didn't do a string addition here, but did the voters calculation before creating the message. Could matter if there's a batch of timeouts happening at the same time (maybe not with group 0, but in the future with multiple groups)

kbr-scylla replied:
> It's just one alloc, I don't think it matters.


Famous last words

kbr-scylla commented on service/storage_service.cc:
```diff
@@ -1105,6 +1105,14 @@ class join_node_rpc_handshaker : public service::group0_handshaker {
+    // _topology_state_machine._topology.normal_nodes to be not empty.
+    // We cannot mark ourselves as dead during this process.
+    // Since there's no chance of raft operations
+    // getting stuck, we don't use raft_timeout{} here.
```
We could use it like a sanity check.

gusev-p replied:
I'm not sure timeout is a good sanity check here. The function `raft_initialize_discovery_leader` is part of the node startup routine, which already includes several valid uses of `raft_timeout{}`. If a timeout occurs here, it would be challenging to differentiate it from other timeouts. Not having a timeout might actually serve as a more effective sanity check - we can't overlook the bootstrapping process getting stuck.

Or maybe we should use `raft_timeout{}`, try/catch our timeout exception and transform it to `on_internal_error` here.

kbr-scylla replied:
> Or maybe we should use raft_timeout{}, try/catch our timeout exception and transform it to on_internal_error here.

Good idea, but we'd have to use something more specific than `runtime_error`

gusev-p replied:
yes

kbr-scylla commented on test/topology_custom/test_raft_no_quorum.py:
```diff
@@ -108,3 +108,44 @@ async def test_quorum_lost_during_node_join(manager: ManagerClient) -> None:
+    logger.info("attempting removenode for the second node")
+    await manager.remove_node(servers[0].server_id, servers[1].server_id,
+                              expected_error="raft operation [read_barrier] timed out, no group0 quorum",
+                              timeout=60)
```
I understand these will timeout much faster than 60s (due to the server-side error injections)?
In that case I wonder why you decided to add `timeout` parameter to these functions instead of still relying on `TOPOLOGY_TIMEOUT` (not saying it's a bad thing, just wondering)

gusev-p replied:
Initially, because of the injections, I used much smaller timeouts, 5 or 10 seconds. This allows you to waste less time in development. But then this started to fail on debug CI and I reverted everything back to the default 60s.

Now I'm starting to think we shouldn't have this timeout arguments at all. We should use one default timeout for everything, say, 5 seconds for 'normal' builds, 60 seconds for debug. Tests should ensure (through injections) that their operations are within that limit.

kbr-scylla replied:
> Initially, because of the injections, I used much smaller timeouts, 5 or 10 seconds. This allows you to waste less time in development. But then this started to fail on debug CI and I reverted everything back to the default 60s.

Wait but doesn't it timeout after 1 second right now? (because of the injections)

The 60s timeout is only for the client side to fail the test earlier **if** the expected server-side timeout doesn't happen, IIUC?

gusev-p replied:
> Wait but doesn't it timeout after 1 second right now? (because of the injections)

Yes, but who knows what happens on CI in debug mode. Actual delays here correlate quite loosely with the expected ones.

> The 60s timeout is only for the client side to fail the test earlier if the expected server-side timeout doesn't happen, IIUC?

Yes, but on CI in debug we should account for a lot of arbitrary delays. I hope 60s is enough for that.

kbr-scylla commented on test/topology_custom/test_raft_no_quorum.py:
```diff
@@ -110,6 +110,55 @@ async def test_quorum_lost_during_node_join(manager: ManagerClient) -> None:
+    await injection_handler.message()
+
+    logger.info("waiting for third node joining process to fail")
+    await third_node_future
```
Could we merge some of these test cases so not to start as many clusters/servers?

gusev-p replied:
I don't like squishing tests into one write-once-understand-never test because of infrastructure constraints. The tests are already quite hard to follow because of injections and asynchrony, it would be even harder if different scenarios started intertwining with each other.

It should be possible to ask the infra for a node and get back one quickly and cheaply. Wouldn't it be more efficient to have some 'reset' api in Scylla, which would drop all the data and recreate all the necessary services?

For now I'll think if I can merge some of the tests without introducing much complexity.

kbr-scylla commented on service/raft/raft_group0.cc:
```diff
@@ -226,13 +226,17 @@ raft_server_for_group raft_group0::create_server_for_group0(raft::group_id gid,
     // initialize the corresponding timer to tick the raft server instance
     auto ticker = std::make_unique<raft_ticker_type>([srv = server.get()] { srv->tick(); });
-
+    lowres_clock::duration timeout = std::chrono::minutes(1);
```
Ok, after reading the rest of the PR it makes sense.
You don't pass `raft_timeout{}` you don't have timeout.
You pass `raft_timeout{}` you get the default (1 minute).
You pass `raft_timeout{X}` you get `X`.

But this deserves a comment to describe the API (for the user) somewhere, perhaps at the definition of `raft_server_with_timeouts`

kbr-scylla commented on service/raft/raft_group_registry.cc:
```diff
@@ -408,6 +413,98 @@ shared_ptr<raft::failure_detector> raft_group_registry::failure_detector() {
+template <std::invocable<abort_source*> Op>
+std::invoke_result_t<Op, abort_source*>
+raft_server_with_timeouts::run_with_timeout(Op&& op, const char* op_name,
+    seastar::abort_source* as, std::optional<raft_timeout> timeout)
```
Now I kinda get it, but seriously `raft_server_with_timeouts` deserves a comment to describe the semantics.

kbr-scylla commented on utils/error_injection.hh:
```diff
@@ -530,6 +560,12 @@ public:
 
+    template <typename T>
+    [[gnu::always_inline]]
+    std::optional<T> inject_parameter(const std::string_view& name) {
```
Probably doesn't matter anyway, this is not used on any hot path at least right now (IIRC from the review)

gusev-p replied:
disassembly of `raft_group_registry::start` for release build:
```
0x00000000046475c5 <+133>: call   0x46444e0 <service::raft_group_registry::init_rpc_verbs()>

   0x00000000046475ca <+138>: cmp    QWORD PTR [rip+0x17b7b46],0x0        # 0x5dff118
   0x00000000046475d2 <+146>: je     0x46475d9 <service::raft_group_registry::start()+153>
   0x00000000046475d4 <+148>: call   0x5dbd270 <TLS init function for utils::error_injection<false>::_local@plt>

   0x00000000046475d9 <+153>: lea    r13,[r15+0x60]
   0x00000000046475dd <+157>: mov    rsi,QWORD PTR [r12+0x50]
   0x00000000046475e2 <+162>: mov    rdx,QWORD PTR [r12+0x60]
   0x00000000046475e7 <+167>: add    rdx,0x8
--Type <RET> for more, q to quit, c to continue without paging--
   0x00000000046475eb <+171>: mov    ecx,0x3b9aca00
   0x00000000046475f0 <+176>: mov    rdi,r13
   0x00000000046475f3 <+179>: call   0x4713ac0 <direct_failure_detector::failure_detector::register_listener(direct_failure_detector::listener&, long)>
```
There is only TLS init function call for `error_injection<false>::_local`, which would be the same for the lambda-based inject functions.

kbr-scylla commented on service/raft/raft_group_registry.cc:
```diff
@@ -408,6 +413,98 @@ shared_ptr<raft::failure_detector> raft_group_registry::failure_detector() {
 raft_group_registry::~raft_group_registry() = default;
 
+namespace {
+    class composite_abort_source {
```
> scylladb/utils/composite_abort_source.hh?

yes

> Should we add a .cc file? 

IMO no need, this is similar to abort_on_expiry.hh (in seastar) which also doesn't have .cc file
Alternatively, utils.cc sounds good too.
Pick your choice

kbr-scylla commented on service/raft/raft_group_registry.cc:
```diff
@@ -408,6 +413,98 @@ shared_ptr<raft::failure_detector> raft_group_registry::failure_detector() {
+        utils::small_vector<abort_source::subscription, 2> _subscriptions;
+    public:
+        void add(abort_source& as) {
+            as.check();
```
Yes, my approach also wouldn't execute the code below (i.e. subscribe), but take alternative path: if `as` is aborted, then immediately `_as.request_abort()` (but don't throw from `add` itself)

kbr-scylla replied:
> This doesn't prevent the code after composite_abort_source::add from executing, throwing exception does.


Yeah, I guess the correct pattern would be then to do `check()` on the composite after `add()`. Throwing from `add()` indeed might be more convenient. Let's say you create a composite and then do a sequence of `add()`s, your way guarantees that the composite is not aborted.

On the other hand, let's say you create composite, `add()`, then yield, and `add()` again. You should still check the composite after yielding because the first `add()`ed abort_source might be aborted by this point...

Perhaps `add()` should also `check()` the composite at the end, if we take your approach?

gusev-p replied:
> > This doesn't prevent the code after composite_abort_source::add from executing, throwing exception does.
> 
> Yeah, I guess the correct pattern would be then to do `check()` on the composite after `add()`. Throwing from `add()` indeed might be more convenient. Let's say you create a composite and then do a sequence of `add()`s, your way guarantees that the composite is not aborted.
> 
> On the other hand, let's say you create composite, `add()`, then yield, and `add()` again. You should still check the composite after yielding because the first `add()`ed abort_source might be aborted by this point...
> 
> Perhaps `add()` should also `check()` the composite at the end, if we take your approach?

Checking on `_as` seems weird to me...  Let's leave it as is for now.

kbr-scylla commented on service/raft/raft_group_registry.cc:
```diff
@@ -408,6 +413,98 @@ shared_ptr<raft::failure_detector> raft_group_registry::failure_detector() {
+        }
+        // TODO: improve error message (check for quorum loss etc)
+        const auto message = ::format("raft operation [{}] timed out", op_name);
+        static logger::rate_limit rate_limit{std::chrono::seconds(1)};
```
> What if we have thousands of raft groups from tablets?

Right, but in that case, the log should definitely not be at WARN level.

> We rethrow only the info that can make sense to the user.

Meh, I don't think there's anything wrong in providing more details in the exception. The user who doesn't understand them can ignore them and focus on the things that they understand. But more advanced person (field engineer or sth) will find the details useful.

gusev-p replied:
This exception may reach client application, I suppose. Seeing C++ file names and line numbers in the log file of some go application (for example) may be confusing.

kbr-scylla replied:
Well. In that spirit, we should avoid mentioning "raft" too, it's internal implementation detail.

kbr-scylla commented on service/raft/raft_group_registry.cc:
```diff
@@ -408,6 +413,98 @@ shared_ptr<raft::failure_detector> raft_group_registry::failure_detector() {
+        static logger::rate_limit rate_limit{std::chrono::seconds(1)};
+        rslog.log(log_level::warn, rate_limit, "{}; timeout requested at [{}], original error {}",
+            message, fmt_loc(timeout->loc), std::current_exception());
+        throw std::runtime_error(message.c_str());
```
Right.

Probably it would be best to use something that's understandable to `cqlsh`. Because, if I remember correctly, `cqlsh` does not print the contents of `runtime_error`, it translates it into vague "server error". But here it would be nice to show that we don't have quorum, for example.

Could we use one of the error types that driver understands?

gusev-p commented:
Now i'm thinking why not just put these three methods with `std::optional<raft_timeout>` arguments right into `raft/server.cc`:
* implementation would be the same, we don't need anything from `raft_registry`, raft::server already has a reference to `failure_detector`;
* we can use single `raft::timeout_error` instead of introducing another class `raft_operation_timeout_error`;
* we can get rid of several entities (`raft_server_with_timeouts` class,  `get_server_with_timeouts`, `group0_with_timeouts` methods), so the code would be simpler;
* `raft_server_with_timeouts` has actually nothing to do with `raft_group_registry`, it lives here just by accident and because it was convenient with the current code structure. 
* the 'keep raft lib minimalistic' argument - ok, but what's the point in that if it leads to the need to introduce unnatural wrappers to get things done? Any user of the base raft library may benefit from the timeouts support, it's not specific to Scylla codebase.

gusev-p commented:
new version:
  * rebased on master, fix conflicts on auth
  * seeds -> servers_add
  * comment raft_server_with_timeouts
  * utils/composite_abort_source.hh
  * raft_server_with_timeout:
    * added thread_local to logger::rate_limit
    * added _group_server.gid to the error message
    * remove 'group0' from the error message (it's general code, not only group0)
    * throw custom error raft_operation_timeout_error instead of std::runtime_error
    * store raft_server_for_group& in the field instead of raft::server + timeout
  * use raft_timeout{} in raft_initialize_discovery_leader
  * merged test_cannot_run_schema_operations into test_cannot_run_operations

@kbr-scylla

gusev-p commented:
[run test_raft_no_quorum.py 100 times in debug](https://jenkins.scylladb.com/job/scylla-master/job/byo/job/byo_build_tests_dtest/2129/)

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7372/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7372/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8499/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/466/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7372/)
#### Build Details:
- Duration: 2 hr 23 min
- Builder: spider3.cloudius-systems.com

gusev-p commented:
> We can't do this, it would lead to double execution. group0_concurrent_modification gives you 100% knowledge that the command had no effect

We would just make another iteration of the topology state machine with the proper read barrier. We would observe a new state and make new decisions, should be no double execution.

gusev-p commented:
> [run test_raft_no_quorum.py 100 times in debug](https://jenkins.scylladb.com/job/scylla-master/job/byo/job/byo_build_tests_dtest/2129/)

[new run with increased timeouts for debug](https://jenkins.scylladb.com/job/scylla-master/job/byo/job/byo_build_tests_dtest/2133/)

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7394/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7394/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/481/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8514/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7394/)
#### Build Details:
- Duration: 2 hr 14 min
- Builder: spider5.cloudius-systems.com

gleb-cloudius commented on service/raft/raft_group_registry.hh:
```diff
@@ -49,6 +51,50 @@ struct raft_server_for_group {
+
+struct raft_timeout {
+    seastar::compat::source_location loc = seastar::compat::source_location::current();
+    std::optional<lowres_clock::duration> value;
```
Better to use time_point here. No need to recalculate duration in case of retrying an operation.

gleb-cloudius commented on service/raft/raft_group_registry.cc:
```diff
@@ -468,8 +471,24 @@ raft_server_with_timeouts::run_with_timeout(Op&& op, const char* op_name,
+            }
+            if (alive_voters_count <= voters_count / 2) {
+                message += ::format(", there is no raft quorum, total voters {}, alive voters {}",
+                    voters_count, alive_voters_count);
```
It would be nice to report the set of alive (or dead) nodes as well.

gusev-p replied:
There could be a lot of them in big clusters. Now this message is thrown to the user, shouldn't contain hundreds of IPs. Maybe we can include several random down nodes for convenience.

gleb-cloudius replied:
Having more than half of the nodes down in a huge cluster should not be a normal event. And eventually we will only have a handful of voters in such situations (we will maintain small subset of voters only. not each node will be a voter), so it we will report dead voters only the list should be relatively short.

gusev-p replied:
done

kbr-scylla commented:
> [new run with increased timeouts for debug](https://jenkins.scylladb.com/job/scylla-master/job/byo/job/byo_build_tests_dtest/2133/)

> UPDATE: passed

ARM debug will probably be even slower than x86 debug.
And then we have to consider that on real CI runs we run many many tests concurrently from different suites.

5 seconds is IMO way too conservative. I would be generous here, why not go as far as 30 seconds

gusev-p commented:
> > [new run with increased timeouts for debug](https://jenkins.scylladb.com/job/scylla-master/job/byo/job/byo_build_tests_dtest/2133/)
> 
> > UPDATE: passed
> 
> ARM debug will probably be even slower than x86 debug. And then we have to consider that on real CI runs we run many many tests concurrently from different suites.
> 
> 5 seconds is IMO way too conservative. I would be generous here, why not go as far as 30 seconds

Each test would take >= 30 seconds to run in debug -- we are waiting for timeout to actually happen.

kbr-scylla commented:
> we are waiting for timeout to actually happen

wait -- so why did you need to increase the timeouts for test to pass? If we wait for it to happen anyway?

gusev-p commented:
> Right. The timeout is modified using error injections -- perhaps we could keep it high by default when we don't expect failures, then decrease it during test when we want a failure to happen? Not sure if it's easy to do though, since the injection is already enabled at startup, dunno what will happen if we try to enable it again in the middle of the test, maybe nothing. Not sure if worth the trouble

This would make the cause of the failure nondeterministic, sometimes we would get a lost quorum in the error message, sometimes we wouldn't.

kbr-scylla commented on utils/error_injection.hh:
```diff
@@ -104,6 +105,14 @@ using error_injection_parameters = std::unordered_map<sstring, sstring>;
  *    Requires func to be a function taking an injection_handler reference and
  *    returning a future<>.
  *    Expected use case: wait for an event from tests.
+ * 5. inject_parameter(name)
```
heads up: in recently merged https://github.com/scylladb/scylladb/pull/17734, `inject_with_handler` was renamed into `inject`, maybe we want to go in this direction here too? I have no opinion

kbr-scylla replied:
(it's currently on the `next` queue, there could also be a conflict)

gusev-p replied:
I don't like when different things are called the same. `inject_with_handler` was behaving quite differently from other `inject`-s. I certainly want a separate name for this new function.

kbr-scylla commented on service/raft/raft_group_registry.hh:
```diff
@@ -49,6 +51,50 @@ struct raft_server_for_group {
+// Use methods of these class if the code is handling a client request
+// and the client expects a potential error. Don't use timeouts for background
+// fibers, such as topology coordinator, since they wouldn't add much value.
+// The only thing the background fiber can do with a timeout is to retry, and
```
@gusev-p @gleb-cloudius I'm actually not sure about that.

For example suppose we're in the middle of bootstrap and coordinator tries to transition from join_group0 to write_both_read_old state.

We lost quorum.

Should the coordinator get stuck there forever? Wouldn't it be better if we had timeout for this group 0 command too? It makes sense to fail the operation instead of sleeping forever.

On the other hand, there is no way to rollback from the operation if we lost quorum. So when coordinator wants to transition back to normal state, then indeed a timeout no longer makes sense.

So I wonder if we should timeout on the forward path of topology sagas, up to the point-of-no-return. After point-of-no-return, or on the abort path, we would not timeout.

WDYT?

gusev-p replied:
> Should the coordinator get stuck there forever? Wouldn't it be better if we had timeout for this group 0 command too? It makes sense to fail the operation instead of sleeping forever.

IIUC, it wouldn't get stuck:
```
void fsm::tick_leader() {
    if (election_elapsed() >= ELECTION_TIMEOUT) {
        // 6.2 Routing requests to the leader
        // A leader in Raft steps down if an election timeout
        // elapses without a successful round of heartbeats to a majority
        // of its cluster; this allows clients to retry their requests
        // with another server.
        return become_follower(server_id{});
    }
```
the leadership will be lost and and topology coordinator fiber will be aborted

kbr-scylla replied:
Right. Do we print any useful message in the log if that happens?
And what happens to a bootstrapping node which gets stuck waiting for response if that happens?

gleb-cloudius replied:
We have a fiber that prints about gaining/loosing leadership in Scylla now. May be it worth adding a log message about coordinator fiber termination as well.

gleb-cloudius commented on service/raft/raft_group_registry.cc:
```diff
@@ -408,6 +424,82 @@ shared_ptr<raft::failure_detector> raft_group_registry::failure_detector() {
+        co_return;
+    }
+    if (!timeout->value) {
+        if (!_group_server.default_op_timeout) {
```
Can't we get rid of this default parameter and instead have a default for `raft_timeout::value`?

gusev-p replied:
The idea was to allow different timeouts for different groups of rafts.

gleb-cloudius replied:
I see. I think group0 is a special one that can use global default timeout. Groups for consistent tables will have different timeouts for different kind of requests, so global setting will be less useful for them. But OK.

gleb-cloudius commented on service/raft/raft_group0.hh:
```diff
@@ -283,6 +283,11 @@ public:
     }
 
+    // Returns a wrapper for the group0_server() with timeouts support.
+    raft_server_with_timeouts group0_server_with_timeouts() {
```
So now we have `raft_server` and `raft_server_with_timeouts`, but `raft_server_with_timeouts` can get `nullopt` as a timeout and also do not have any timeouts. Why not have only one `raft_server`  that gets timeout and if no timeout is needed the caller will pass `nullopt` as a timeout parameter?

gusev-p replied:
> Why not have only one raft_server that gets timeout and if no timeout is needed the caller will pass nullopt as a timeout parameter?

don't need timeout -> pass `std::nullopt` as raft_timeout parameter
need specific timeout -> pass `raft_timeout{<specific_timeout_value>}` (not used right now)
need default for group timeout -> pass `raft_timeout{}`

gusev-p commented on service/raft/raft_group0_client.hh:
```diff
@@ -107,7 +107,7 @@ public:
     future<> init();
 
-    future<> add_entry(group0_command group0_cmd, group0_guard guard, seastar::abort_source* as);
+    future<> add_entry(group0_command group0_cmd, group0_guard guard, seastar::abort_source* as, std::optional<raft_timeout> timeout = std::nullopt);
```
> If we want a default, I think raft_timeout{} is a better one.

Default value usually disables behaviour. It's more natural to expect that explicitly passed parameter adds behaviour, not removes it.

> I prefer debugging an explicit unexpected failure, than debugging a hang. If we have a failure we usually instantly know where to look. If we have a hang... it's much harder.

I strongly agree with this. Why not then take this a step further and declare 1 minute (or some other value) a group0 default operation timeout? We can still provide raft_timeout parameter with different semantics - if it's not null, its value field overrides the group level timeout parameter, if it's null - no timeout is applied.

kbr-scylla replied:
> Default value usually disables behaviour. It's more natural to expect that explicitly passed parameter adds behaviour, not removes it.

Philosophy. What's behavior here? Is waiting forever a behavior? Is waiting a limited amount of time a behavior? Which behavior is more of a behavior?

> I strongly agree with this. Why not then take this a step further and declare 1 minute (or some other value) a group0 default operation timeout? We can still provide raft_timeout parameter with different semantics - if it's not null, its value field overrides the group level timeout parameter, if it's null - no timeout is applied.

Didn't you just describe exactly what the PR already does?

The problem is if you call `add_entry` and forget to make a decision, i.e. don't pass `raft_timeout`, `nullopt` will be used, meaning that we will not have a timeout at all.

So the question is which outcome is better if the caller of `add_entry` forgets to make a decision.
- if the correct decision would be to wait forever, but instead we will timeout after 1 minute
- or if the correct decision would be to timeout, but instead we wait forever (hang)?

I think out of the two undesirable outcomes, the first one is better -- because it's much easier to debug.

gusev-p replied:
> Didn't you just describe exactly what the PR already does?

Nope, in the current PR version not passed raft_timeout parameter means don't apply timeout. I propose to change this to 'I don't care, use group-wide default'. If someone cares about timeouts, they can provide raft_timeout with explicit value, null for no timeout, time_point for specific timeout. This is currently not needed though if we take my approach (see below).

> I think out of the two undesirable outcomes, the first one is better -- because it's much easier to debug.

I agree.

We are basically talking about the same, the only question is what to do with background fibers. In any case, we would need to review all group0 usages in all of them. In your approach we would need to add `raft_timeout{}` to group0 calls, in my approach - add the `try/catch` to the top loop. I like my approach more because it doesn't require passing `raft_timeout{}` parameter down the call chains of the methods, which are used both in fibers and in user context (like in [auth commit](https://github.com/scylladb/scylladb/pull/17590/commits/da9d2b6ea0baedd5031756925f462fbbad5b9a5e), and there are several other such cases).

kbr-scylla replied:
> in my approach - add the try/catch to the top loop

So you're proposing to **always** have a timeout, and then deal with it with a try/catch + retry

My proposal is to have a timeout **by default**, but give the option to not have a timeout when needed (in background fibers), and pass it explicitly there

kbr-scylla replied:
>  In your approach we would need to add raft_timeout{} to group0 calls


Ah, you meant those group 0 calls that are being used in both contexts where we want and don't want a timeout

But I think there are fewer cases where we don't want timeout

gusev-p replied:
> In my approach we would not need to add raft_timeout{}. Because it would be the default
> I.e. change this:

I mean add std::nullopt to the call chains originating from background fibers.

kbr-scylla replied:
> have only top level timeout for entire operations (bootstrap a node, removenode, decomission etc).

We cannot do that. Those operations have unbounded runtime, based on the size of cluster / size of data etc.

It makes sense to have timeouts only for operations that should have a bound like schema changes or auth changes

kbr-scylla replied:
> I mean add std::nullopt to the call chains originating from background fibers.

Yes yes

but I actually think there are not  that many of them

For example in topology_coordinator there is... one :)

What are the others?

gusev-p replied:
> We cannot do that. Those operations have unbounded runtime, based on the size of cluster / size of data etc.
> 
> It makes sense to have timeouts only for operations that should have a bound like schema changes or auth changes

True.

kbr-scylla replied:
> For example in topology_coordinator there is... one :)

Ok sorry, I see two:
```cpp
    future<group0_guard> start_operation() {
        auto guard = co_await _group0.client().start_operation(&_as);
```

and
```cpp
    future<> update_topology_state(
            group0_guard guard, std::vector<canonical_mutation>&& updates, const sstring& reason) {
        try {
            rtlogger.info("updating topology state: {}", reason);
            rtlogger.trace("update_topology_state mutations: {}", updates);
            topology_change change{std::move(updates)};
            group0_command g0_cmd = _group0.client().prepare_command(std::move(change), guard, reason);
            co_await _group0.client().add_entry(std::move(g0_cmd), std::move(guard), &_as);
```

kbr-scylla replied:
I also think it's fine to have a timeout in the upgrade procedure fiber.

It is being triggered by a POST and waited on by a GET, so while it is not strictly "synchronous", the documented procedure is for the admin to monitor its progress by GET.

kbr-scylla replied:
> Nothing stops new ones from appearing.

And the same is true in the other direction i.e. those that should have a timeout but somebody will forget to pass it.

Again we're back to the dilemma presented above, what is worse -- new call will appear inside topo coordinator code that will timeout, OR new call will appear outside topo coordinator that will not timeout when it should

gusev-p replied:
> Again we're back to the dilemma presented above, what is worse -- new call will appear inside topo coordinator code that will timeout, OR new call will appear outside topo coordinator that will not timeout when it should

No, we're not. With both our approaches user will get timeout by default, but in my case they don't need additional plumbing for that. The plumbing, which adds more places where the wrong parameter could be passed.

kbr-scylla replied:
Then I don't understand. In your approach, if user wants to specify "no timeout", how do they do this? Or do you want to take away this possibility? (So it's not "timeout by default", it is "timeout always") And how do they specify a different timeout than the default?

gusev-p replied:
* My main point is that I want to reduce the amount of plumbing in the code;
  * I don’t want to introduce long chains where we need to pass raft_timeout{} from one function to another;
  * Simple solution - get rid of the reason why we need raft_timeout{} in one place but not in another
  * This means all group0 methods should have a default timeout; background fibers catch raft_operation_timeout_exception and retry at the top level
  * In the PR I proposed to introduce default timeout for raft_server_for_group, but it’s an implementation detail and not particularly important. We can introduce some global group0_timeout = 1 minute and use it in those places that communicate directly with raft_group_registry for group0 (mainly raft_group0_client + a few other)
  * The app level code (storage_service, migration_manager) won’t have to care about timeouts, except in try/catch for background fibers. We even don’t need raft_timeout parameter in raft_group0_client methods, they just always pass group0_timeout to the underlying raft_group0_registry api-s.

kbr-scylla replied:
As discussed on yesterday's daily -- let's proceed with the current approach.

gusev-p replied:
ok

gusev-p commented:
new version:
  * report dead nodes in `raft_operation_timeout_error`

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:x: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7568/)
#### Build Details:
- Duration: 13 min
- Builder: spider1.cloudius-systems.com

gusev-p commented:
new version:
  * rebased on master

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7571/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7571/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/579/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8609/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7571/)
#### Build Details:
- Duration: 2 hr 9 min
- Builder: spider6.cloudius-systems.com

gusev-p commented:
@kbr-scylla

kbr-scylla commented:
Dequeued, breaks on ARM debug

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7629/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7629/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/627/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8659/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7629/)
#### Build Details:
- Duration: 2 hr 12 min
- Builder: spider3.cloudius-systems.com

gusev-p commented:
@kbr-scylla
